<a href="https://colab.research.google.com/github/monalisamagar/restaurant-data-analytics/blob/main/ITS69304_0362948_INDIVASGNMT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# first look

In [ ]:
#@title pandas
import pandas as pd #library data manipulation
df = pd.read_excel('/content/5501-5758.xlsx')

In [ ]:
#@title warning
import warnings
warnings.filterwarnings("ignore")

In [ ]:
#@title set option
pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', 4)

:10

yyyy-mm-dd hrs-min-sec

In [ ]:
#@title display original dataset
display(df)

S. NO.    REGISTRATION DATE  \
0      5501  2070-12-20 00:00:00   
1      5502  2070-12-23 00:00:00   
2      5503  2070-12-24 00:00:00   
3      5504  2070-12-24 00:00:00   
4      5505  2070-12-26 00:00:00   
..      ...                  ...   
253    5754  2071-06-08 00:00:00   
254    5755  2071-06-08 00:00:00   
255    5756  2071-06-10 00:00:00   
256    5757  2071-06-10 00:00:00   
257    5758  2071-06-10 00:00:00   

                                     INDUSTRY NAME   DISTRICT  TOTAL CAPITAL  \
0    DENG FENG LAI HOTEL AND RESTAURANT  PVT. LTD.      KASKI       10000000   
1                 CHINA MOON RESTAURENT  PVT. LTD.      KASKI        5000000   
2                           GANSU ENERGY PVT. LTD.  KATHMANDU       30000000   
3       PASTAPUR ITALIAN FOOD RESTAURANT PVT. LTD.   LALITPUR        5000000   
4      KATHMANDU CHINESE MEE RESTAURANT  PVT. LTD.  KATHMANDU       25000000   
..                                             ...        ...            ...   
253   C.N. YUAN XIN CONSTRUCTION COMPANY PVT. LTD.   LALITPUR       30000000   
254        AKLEKUNDA HYDROPOWER\nCOMPANY PVT. LTD.    NUWAKOT      160200000   
255           STARKEY LABORATORIES NEPAL PVT. LTD.   LALITPUR       14046368   
256       LU LU TONG INTERNATIONAL CARGO PVT. LTD.  KATHMANDU       25000000   
257       YE CHEN NOOD LES MANUFACTURING PVT. LTD.  KATHMANDU       29000000   

     FIXED CAPITAL  WORKING CAPITAL  EMPLOYMENT       CATEGORY  SCALE  \
0          7500000          2500000          41        TOURISM  SMALL   
1          3500000          1500000          25        TOURISM  SMALL   
2          9100000         20900000          33   ENERGY BASED  SMALL   
3          4100000           900000          14        TOURISM  SMALL   
4         21000000          4000000          40        TOURISM  SMALL   
..             ...              ...         ...            ...    ...   
253       19900000         10100000          81        SERVICE  SMALL   
254      158000000          2200000          20   ENERGY BASED  LARGE   
255        5300000          8746368           8  MANUFACTURING  SMALL   
256       20000000          5000000          32        SERVICE  SMALL   
257       20000000          9000000          30  MANUFACTURING  SMALL   

       POWER % OF INVESTMENT  
0     75 KVA  Foreign - 100%  
1     10 KVA  Foreign - 100%  
2     50 KVA  Foreign - 100%  
3     15 KVA  Foreign - 100%  
4     20 KVA  Foreign - 100%  
..       ...             ...  
253  100 KVA  Foreign - 100%  
254   20 KVA    Local - 100%  
255    5 KVA  Foreign - 100%  
256   10 KVA  Foreign - 100%  
257   75 KVA  Foreign - 100%  

[258 rows x 12 columns]

In [ ]:
#@title copy df
df1 = df.copy()

In [ ]:
#@title info
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 258 entries, 0 to 257
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   S. NO.             258 non-null    int64 
 1   REGISTRATION DATE  258 non-null    object
 2   INDUSTRY NAME      257 non-null    object
 3   DISTRICT           257 non-null    object
 4   TOTAL CAPITAL      258 non-null    int64 
 5   FIXED CAPITAL      258 non-null    int64 
 6   WORKING CAPITAL    258 non-null    int64 
 7   EMPLOYMENT         258 non-null    int64 
 8   CATEGORY           258 non-null    object
 9   SCALE              258 non-null    object
 10  POWER              258 non-null    object
 11  % OF INVESTMENT    258 non-null    object
dtypes: int64(5), object(7)
memory usage: 24.3+ KB


In [ ]:
#@title drop it NAN
df1 = df1.dropna()

In [ ]:
#@title describe df1 only numerical
display(df1.describe())

S. NO.  TOTAL CAPITAL  FIXED CAPITAL  WORKING CAPITAL  EMPLOYMENT
count   256.000000   2.560000e+02   2.560000e+02     2.560000e+02  256.000000
mean   5629.585938   2.979633e+08   2.832750e+08     1.468824e+07   49.617188
std      74.845727   2.261194e+09   2.241519e+09     3.527673e+07   69.794558
min    5501.000000   8.000000e+05   6.000000e+05     2.000000e+05    8.000000
25%    5564.750000   1.000000e+07   8.500000e+06     1.500000e+06   24.750000
50%    5629.500000   2.500000e+07   2.025000e+07     3.000000e+06   32.000000
75%    5694.250000   4.910000e+07   3.700000e+07     9.000000e+06   50.000000
max    5758.000000   3.500000e+10   3.475000e+10     2.500000e+08  925.000000

In [ ]:
#@title scale value count
display(df1['SCALE'].value_counts())

SCALE
SMALL     181
LARGE      42
MEDIUM     33
Name: count, dtype: int64

In [ ]:
#@title category value count
display(df1['CATEGORY'].value_counts())

CATEGORY
SERVICE                    77
TOURISM                    72
MANUFACTURING              53
AGR0 AND FORESTRY BASED    32
ENERGY BASED               17
MINERAL                     4
CONSTRUCTION                1
Name: count, dtype: int64

In [ ]:
#@title dsitrict nunique and value counts
display(df1['DISTRICT'].value_counts())
print(df1['DISTRICT'].nunique())

DISTRICT
KATHMANDU         109
LALITPUR           54
KASKI              20
MORANG              7
KAVRE               7
PARSA               6
SUNSARI             5
RUPANDEHI           4
DANG                4
JHAPA               4
GORKHA              3
CHITWAN             3
BHAKTAPUR           2
SANKHUWASABHA       2
NUWAKOT             2
SINDHUPALCHOWK      2
TANAHU              2
BARA                2
RAMECHHAP           1
ARGHAKHACHI         1
KAILALI             1
MYAGD I             1
DHADING             1
PARBAT              1
RASUWA              1
ILAM                1
TAPLEJUNG           1
SYANGJA             1
DHANUSHA            1
DOLKHA              1
BAGLUNG             1
RAUTAHAT            1
BANKE               1
BARDIYA             1
MUSTANG             1
NAWALPARASI         1
Name: count, dtype: int64

36


In [ ]:
#@title % of investment nunique and calue count
print(df1['% OF INVESTMENT'].nunique())
display(df1['% OF INVESTMENT'].value_counts())

18


% OF INVESTMENT
Foreign - 100%                      168
Local  -  100%                       62
Local - 50%,\nForeign - 50%           4
Local - 25%,\nForeign - 75%           4
Local - 12%,\nForeign - 88%           2
Local - 33.34%, Foreign - 66.66%      2
Local - 16.67%, Foreign - 83.33%      2
Local - 60%,\nForeign - 40%           2
Local - 5.66%, Foreign - 94.34%       1
Local - 57%,\nForeign - 43%           1
Local - 9.1%,\nForeign - 90.9%        1
Local - 10%,\nForeign - 90%           1
Local - 6.67%, Foreign - 93.33%       1
Local - 58.28%, Foreign - 41.72%      1
Local - 16.66%, Foreign - 83.34%      1
Local - 66.67%, Foreign - 33.33%      1
Local - 34%,\nForeign - 66%           1
Local - 100%                          1
Name: count, dtype: int64

# clean

In [ ]:
#@title copy to clean
df2=df1.copy()

In [ ]:
#@title check data type
df2.dtypes

S. NO.                int64
REGISTRATION DATE    object
INDUSTRY NAME        object
DISTRICT             object
TOTAL CAPITAL         int64
FIXED CAPITAL         int64
WORKING CAPITAL       int64
EMPLOYMENT            int64
CATEGORY             object
SCALE                object
POWER                object
% OF INVESTMENT      object
dtype: object

In [ ]:
#@title drop SNo
df2 = df2.drop(['S. NO.'], axis=1)

In [ ]:
#@title fix date column convert to ad and extract year and month

# convert  to a string
df2['REGISTRATION DATE'] = df2['REGISTRATION DATE'].astype(str)

# format is YYYY-MM-DD.
df2['MONTH'] = df2['REGISTRATION DATE'].str[5:7]

# convert only date part
df2['REGISTRATION DATE'] = df2['REGISTRATION DATE'].str[:10]

# convert Bikram Sambat to Anno Domini
def bs_to_ad(bs_year):
    ad_year = bs_year - 57
    return ad_year

# apply the conversion to the year column
df2['YEAR'] = df2['REGISTRATION DATE'].str[:4].apply(lambda x: bs_to_ad(int(x)))

# drop original
df2 = df2.drop('REGISTRATION DATE', axis=1)

In [ ]:
#@title regular expression
import re

In [ ]:
#@title turn % of investment to standard format
df2['% OF INVESTMENT'] = df2['% OF INVESTMENT'].str.replace('\n|0Zo', ' ', regex=False)
df2['% OF INVESTMENT'] = df2['% OF INVESTMENT'].str.findall(r'(Local\s*-\s*[\d.,]+%|Foreign\s*-\s*[\d.,]+%)').apply(lambda x: ', '.join(x) if x else '')
df2['% OF INVESTMENT'] = df2['% OF INVESTMENT'].str.replace(r'\s+', ' ', regex=True).str.strip()

In [ ]:
# Check for null values in '% OF INVESTMENT'
print(df2['% OF INVESTMENT'].isnull().sum())

# Check for empty strings in '% OF INVESTMENT'
print((df2['% OF INVESTMENT'] == '').sum())

# Check for NaN values in '% OF INVESTMENT'
print(df2['% OF INVESTMENT'].isna().sum())

0
0
0


In [ ]:
#@title clean power only keep numeric
df2['POWER'] = df2['POWER'].str.replace('\n', ' ', regex=False).str.replace(r'\s+', ' ', regex=True).str.replace('KVA', '', regex=False).str.strip()

In [ ]:
df2['POWER'] = pd.to_numeric(df2['POWER'], errors='coerce')

In [ ]:
#@title display clean dataset
display(df2)

INDUSTRY NAME   DISTRICT  TOTAL CAPITAL  \
0    DENG FENG LAI HOTEL AND RESTAURANT  PVT. LTD.      KASKI       10000000   
1                 CHINA MOON RESTAURENT  PVT. LTD.      KASKI        5000000   
2                           GANSU ENERGY PVT. LTD.  KATHMANDU       30000000   
3       PASTAPUR ITALIAN FOOD RESTAURANT PVT. LTD.   LALITPUR        5000000   
4      KATHMANDU CHINESE MEE RESTAURANT  PVT. LTD.  KATHMANDU       25000000   
..                                             ...        ...            ...   
253   C.N. YUAN XIN CONSTRUCTION COMPANY PVT. LTD.   LALITPUR       30000000   
254        AKLEKUNDA HYDROPOWER\nCOMPANY PVT. LTD.    NUWAKOT      160200000   
255           STARKEY LABORATORIES NEPAL PVT. LTD.   LALITPUR       14046368   
256       LU LU TONG INTERNATIONAL CARGO PVT. LTD.  KATHMANDU       25000000   
257       YE CHEN NOOD LES MANUFACTURING PVT. LTD.  KATHMANDU       29000000   

     FIXED CAPITAL  WORKING CAPITAL  EMPLOYMENT       CATEGORY  SCALE  POWER  \
0          7500000          2500000          41        TOURISM  SMALL   75.0   
1          3500000          1500000          25        TOURISM  SMALL   10.0   
2          9100000         20900000          33   ENERGY BASED  SMALL   50.0   
3          4100000           900000          14        TOURISM  SMALL   15.0   
4         21000000          4000000          40        TOURISM  SMALL   20.0   
..             ...              ...         ...            ...    ...    ...   
253       19900000         10100000          81        SERVICE  SMALL  100.0   
254      158000000          2200000          20   ENERGY BASED  LARGE   20.0   
255        5300000          8746368           8  MANUFACTURING  SMALL    5.0   
256       20000000          5000000          32        SERVICE  SMALL   10.0   
257       20000000          9000000          30  MANUFACTURING  SMALL   75.0   

    % OF INVESTMENT MONTH  YEAR  
0    Foreign - 100%    12  2013  
1    Foreign - 100%    12  2013  
2    Foreign - 100%    12  2013  
3    Foreign - 100%    12  2013  
4    Foreign - 100%    12  2013  
..              ...   ...   ...  
253  Foreign - 100%    06  2014  
254    Local - 100%    06  2014  
255  Foreign - 100%    06  2014  
256  Foreign - 100%    06  2014  
257  Foreign - 100%    06  2014  

[256 rows x 12 columns]

# feature engineering

In [ ]:
#@title view list of columns
columns = df2.columns.tolist()
print("Columns:", columns)

Columns: ['INDUSTRY NAME', 'DISTRICT', 'TOTAL CAPITAL', 'FIXED CAPITAL', 'WORKING CAPITAL', 'EMPLOYMENT', 'CATEGORY', 'SCALE', 'POWER', '% OF INVESTMENT', 'MONTH', 'YEAR']


In [ ]:
#@title district and region dictionary
province_district = {
    'KOSHI': ['TAPLEJUNG', 'SANKHUWASABHA', 'SOLUKHUMBU', 'UDAYAPUR',
              'PANCHTHAR', 'ILAM', 'TERHATHUM', 'DHANKUTA', 'BHOJPUR',
              'KHOTANG', 'OKHALDHUNGA', 'JHAPA', 'MORANG', 'SUNSARI'],
    'MADHESH': ['MAHOTTARI', 'RAUTAHAT', 'DHANUSHA', 'SIRAHA',
                'BARA', 'SARLAHI', 'PARSA', 'SAPTARI'],
    'BAGMATI': ['DOLKHA', 'SINDHUPALCHOWK', 'RASUWA', 'MAKWANPUR',
                'BHAKTAPUR', 'LALITPUR', 'KATHMANDU', 'NUWAKOT',
                'RAMECHHAP', 'KAVRE', 'DHADING', 'SINDHULI', 'CHITWAN'],
    'GANDAKI': ['MANANG', 'MUSTANG', 'PARBAT', 'SYANGJA',
                'TANAHU', 'LAMJUNG', 'BAGLUNG', 'KASKI',
                'MYAGD I', 'GORKHA', 'NAWALPARASI'],
    'LUMBINI': ['PALPA', 'ARGHAKHACHI', 'RUKUM', 'GULMI',
                'PYUTHAN', 'ROLPA', 'RUPANDEHI', 'KAPILBASTU',
                'DANG', 'BANKE', 'BARDIYA'],
    'KARNALI': ['DOLPA', 'MUGU', 'HUMLA', 'JUMLA', 'SALYAN',
                'JAJARKOT', 'DAILEKH', 'SURKHET', 'KALIKOT'],
    'SUDUR-PASCHIM': ['BAJURA', 'BAJHANG', 'DARCHULA', 'ACHHAM',
                      'DOTI', 'BAITADI', 'KAILALI', 'KANCHANPUR']
}

In [ ]:
#@title apply function to map district to province
def map_district_to_province(district):
    for province, districts in province_district.items():
        if district in districts:
            return province
    return None

df2['PROVINCE'] = df2['DISTRICT'].apply(map_district_to_province)

In [ ]:
#@title make a copy
df3 = df2.copy(deep=True)

# encoding and outlier

In [ ]:
#@title apply frequency encoding to category
district_counts = df3['CATEGORY'].value_counts(normalize=True)
df3['CATEGORY_FREQ'] = df3['CATEGORY'].map(district_counts)
display(df3[['CATEGORY','CATEGORY_FREQ']].value_counts())

CATEGORY                 CATEGORY_FREQ
SERVICE                  0.300781         77
TOURISM                  0.281250         72
MANUFACTURING            0.207031         53
AGR0 AND FORESTRY BASED  0.125000         32
ENERGY BASED             0.066406         17
MINERAL                  0.015625          4
CONSTRUCTION             0.003906          1
Name: count, dtype: int64

In [ ]:
#@title reduceing district to province
display(df3['PROVINCE'].value_counts())

PROVINCE
BAGMATI          183
GANDAKI           31
KOSHI             20
LUMBINI           11
MADHESH           10
SUDUR-PASCHIM      1
Name: count, dtype: int64

Reasoning:

Dimensionality Reduction: Reducing the number of unique values in a categorical feature like 'DISTRICT' can help reduce the dimensionality of your data. This can potentially lead to improved performance and efficiency in clustering and machine learning algorithms, especially if 'DISTRICT' is used as a feature in your model.
Generalization: By grouping districts into provinces, you are essentially creating a higher-level geographical category. This can help the model capture broader patterns and relationships rather than focusing on very specific district-level differences. This can improve the model's ability to generalize to unseen data.
Avoiding Overfitting: With fewer unique categories, the model is less likely to overfit to specific district-level patterns, which can make it more robust and applicable to a wider range of data.
Information Loss: Grouping districts into provinces does come with a trade-off – you might lose some information that was specific to individual districts. If this district-level information is crucial for the model's performance, the change might not be beneficial.
Recommendation:

In your case, changing 'DISTRICT' to 'PROVINCE' is likely to help with clustering and machine learning. Here's why:

It significantly reduces the number of unique values (from 77 to 7), addressing the high cardinality issue.
It helps with generalization, preventing the model from focusing on small, district-specific variations.
It might reduce overfitting and improve model efficiency.

In [ ]:
#@title import label encoding for scale

from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df3['SCALE_encode'] = le.fit_transform(df3['SCALE'])
df3['SCALE_encode'] = df3['SCALE_encode'] + 1

display(df3[['SCALE','SCALE_encode']].value_counts())

SCALE   SCALE_encode
SMALL   3               181
LARGE   1                42
MEDIUM  2                33
Name: count, dtype: int64

In [ ]:
#@title import one hot encoding for province
from sklearn.preprocessing import OneHotEncoder

ohe = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

ohe.fit(df3[['PROVINCE']])

encoded_data = ohe.transform(df3[['PROVINCE']])

feature_names = ohe.get_feature_names_out(['PROVINCE'])
for i, feature_name in enumerate(feature_names):
    df3[feature_name] = encoded_data[:, i]

In [ ]:
#@title display encoded dataframe
display(df3)

INDUSTRY NAME   DISTRICT  TOTAL CAPITAL  \
0    DENG FENG LAI HOTEL AND RESTAURANT  PVT. LTD.      KASKI       10000000   
1                 CHINA MOON RESTAURENT  PVT. LTD.      KASKI        5000000   
2                           GANSU ENERGY PVT. LTD.  KATHMANDU       30000000   
3       PASTAPUR ITALIAN FOOD RESTAURANT PVT. LTD.   LALITPUR        5000000   
4      KATHMANDU CHINESE MEE RESTAURANT  PVT. LTD.  KATHMANDU       25000000   
..                                             ...        ...            ...   
253   C.N. YUAN XIN CONSTRUCTION COMPANY PVT. LTD.   LALITPUR       30000000   
254        AKLEKUNDA HYDROPOWER\nCOMPANY PVT. LTD.    NUWAKOT      160200000   
255           STARKEY LABORATORIES NEPAL PVT. LTD.   LALITPUR       14046368   
256       LU LU TONG INTERNATIONAL CARGO PVT. LTD.  KATHMANDU       25000000   
257       YE CHEN NOOD LES MANUFACTURING PVT. LTD.  KATHMANDU       29000000   

     FIXED CAPITAL  WORKING CAPITAL  EMPLOYMENT       CATEGORY  SCALE  POWER  \
0          7500000          2500000          41        TOURISM  SMALL   75.0   
1          3500000          1500000          25        TOURISM  SMALL   10.0   
2          9100000         20900000          33   ENERGY BASED  SMALL   50.0   
3          4100000           900000          14        TOURISM  SMALL   15.0   
4         21000000          4000000          40        TOURISM  SMALL   20.0   
..             ...              ...         ...            ...    ...    ...   
253       19900000         10100000          81        SERVICE  SMALL  100.0   
254      158000000          2200000          20   ENERGY BASED  LARGE   20.0   
255        5300000          8746368           8  MANUFACTURING  SMALL    5.0   
256       20000000          5000000          32        SERVICE  SMALL   10.0   
257       20000000          9000000          30  MANUFACTURING  SMALL   75.0   

    % OF INVESTMENT MONTH  YEAR PROVINCE  CATEGORY_FREQ  SCALE_encode  \
0    Foreign - 100%    12  2013  GANDAKI       0.281250             3   
1    Foreign - 100%    12  2013  GANDAKI       0.281250             3   
2    Foreign - 100%    12  2013  BAGMATI       0.066406             3   
3    Foreign - 100%    12  2013  BAGMATI       0.281250             3   
4    Foreign - 100%    12  2013  BAGMATI       0.281250             3   
..              ...   ...   ...      ...            ...           ...   
253  Foreign - 100%    06  2014  BAGMATI       0.300781             3   
254    Local - 100%    06  2014  BAGMATI       0.066406             1   
255  Foreign - 100%    06  2014  BAGMATI       0.207031             3   
256  Foreign - 100%    06  2014  BAGMATI       0.300781             3   
257  Foreign - 100%    06  2014  BAGMATI       0.207031             3   

     PROVINCE_BAGMATI  PROVINCE_GANDAKI  PROVINCE_KOSHI  PROVINCE_LUMBINI  \
0                 0.0               1.0             0.0               0.0   
1                 0.0               1.0             0.0               0.0   
2                 1.0               0.0             0.0               0.0   
3                 1.0               0.0             0.0               0.0   
4                 1.0               0.0             0.0               0.0   
..                ...               ...             ...               ...   
253               1.0               0.0             0.0               0.0   
254               1.0               0.0             0.0               0.0   
255               1.0               0.0             0.0               0.0   
256               1.0               0.0             0.0               0.0   
257               1.0               0.0             0.0               0.0   

     PROVINCE_MADHESH  PROVINCE_SUDUR-PASCHIM  
0                 0.0                     0.0  
1                 0.0                     0.0  
2                 0.0                     0.0  
3                 0.0                     0.0  
4                 0.0                     0.0  
..                ...    

In [ ]:
#@title info after encode
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 256 entries, 0 to 257
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   INDUSTRY NAME           256 non-null    object 
 1   DISTRICT                256 non-null    object 
 2   TOTAL CAPITAL           256 non-null    int64  
 3   FIXED CAPITAL           256 non-null    int64  
 4   WORKING CAPITAL         256 non-null    int64  
 5   EMPLOYMENT              256 non-null    int64  
 6   CATEGORY                256 non-null    object 
 7   SCALE                   256 non-null    object 
 8   POWER                   246 non-null    float64
 9   % OF INVESTMENT         256 non-null    object 
 10  MONTH                   256 non-null    object 
 11  YEAR                    256 non-null    int64  
 12  PROVINCE                256 non-null    object 
 13  CATEGORY_FREQ           256 non-null    float64
 14  SCALE_encode            256 non-null    int64  

In [ ]:
#@title display top values from each columns
for column in columns:
    print(f"\nColumn: {column}")
    print(df3[column].values[:4])


Column: INDUSTRY NAME
['DENG FENG LAI HOTEL AND RESTAURANT  PVT. LTD.'
 'CHINA MOON RESTAURENT  PVT. LTD.' 'GANSU ENERGY PVT. LTD.'
 'PASTAPUR ITALIAN FOOD RESTAURANT PVT. LTD.']

Column: DISTRICT
['KASKI' 'KASKI' 'KATHMANDU' 'LALITPUR']

Column: TOTAL CAPITAL
[10000000  5000000 30000000  5000000]

Column: FIXED CAPITAL
[7500000 3500000 9100000 4100000]

Column: WORKING CAPITAL
[ 2500000  1500000 20900000   900000]

Column: EMPLOYMENT
[41 25 33 14]

Column: CATEGORY
['TOURISM' 'TOURISM' 'ENERGY BASED' 'TOURISM']

Column: SCALE
['SMALL' 'SMALL' 'SMALL' 'SMALL']

Column: POWER
[75. 10. 50. 15.]

Column: % OF INVESTMENT
['Foreign - 100%' 'Foreign - 100%' 'Foreign - 100%' 'Foreign - 100%']

Column: MONTH
['12' '12' '12' '12']

Column: YEAR
[2013 2013 2013 2013]


In [ ]:
#@title check for any null value after mapping
display(df3[df3['PROVINCE'].isnull()])

Empty DataFrame
Columns: [INDUSTRY NAME, DISTRICT, TOTAL CAPITAL, FIXED CAPITAL, WORKING CAPITAL, EMPLOYMENT, CATEGORY, SCALE, POWER, % OF INVESTMENT, MONTH, YEAR, PROVINCE, CATEGORY_FREQ, SCALE_encode, PROVINCE_BAGMATI, PROVINCE_GANDAKI, PROVINCE_KOSHI, PROVINCE_LUMBINI, PROVINCE_MADHESH, PROVINCE_SUDUR-PASCHIM]
Index: []

In [ ]:
df4=df3.copy()

# model

In [ ]:
#@title import scipy
from scipy.stats import iqr

In [ ]:
#@title apply interquartile range with scipy in total capital. fixed capital and working capital

# calculate IQR for specified columns
for col in ['TOTAL CAPITAL', 'FIXED CAPITAL', 'WORKING CAPITAL']:
    q1 = df4[col].quantile(0.25)
    q3 = df4[col].quantile(0.75)
    iqr_val = iqr(df4[col])
    print(f"IQR for {col}: {iqr_val}\n")

    # filtering out outliers
    upper_bound = q3 + 1.5 * iqr_val
    lower_bound = q1 - 1.5 * iqr_val
    df4 = df4[(df4[col] >= lower_bound) & (df4[col] <= upper_bound)]

IQR for TOTAL CAPITAL: 39100000.0

IQR for FIXED CAPITAL: 15450000.0

IQR for WORKING CAPITAL: 2620000.0



In [ ]:
#@ title import plotly
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [ ]:
#@title create box plots before and after IQR treatment
fig_before = px.box(df2, y=['TOTAL CAPITAL', 'FIXED CAPITAL', 'WORKING CAPITAL'],
                    title='Box Plots Before IQR Treatment')
fig_before.show()

fig_after = px.box(df4, y=['TOTAL CAPITAL', 'FIXED CAPITAL', 'WORKING CAPITAL'],
                   title='Box Plots After IQR Treatment')
fig_after.show()

In [ ]:
#@title apply standard scaler
from sklearn.preprocessing import StandardScaler

# intitailize
scaler = StandardScaler()

#  columns to standardize
columns_to_standardize = ['TOTAL CAPITAL', 'FIXED CAPITAL', 'WORKING CAPITAL']

# standardized columns into new columns.
for col in columns_to_standardize:
  # fit scaler
  scaler.fit(df4[[col]])

  # new column with the standardized values
  df4[col + '_standardized'] = scaler.transform(df4[[col]])

In [ ]:
def split_investment(investment_str):
    local_pct = 0
    foreign_pct = 0
    try:
      parts = investment_str.split(',')
      for part in parts:
          if 'Local' in part:
              local_pct = float(part.split('-')[1].replace('%', '').strip())
          elif 'Foreign' in part:
              foreign_pct = float(part.split('-')[1].replace('%', '').strip())
    except:
        print(f"Error processing: {investment_str}")
        return local_pct, foreign_pct
    return local_pct, foreign_pct

df4[['Local_Investment', 'Foreign_Investment']] = df2['% OF INVESTMENT'].apply(lambda x: pd.Series(split_investment(x)))

In [ ]:
#@title unique values for Local and Foreign Investment
print("Unique Local Investment values:")
display(df4['Local_Investment'].value_counts())
print("\nUnique Foreign Investment values:")
display(df4['Foreign_Investment'].value_counts())

Unique Local Investment values:


Local_Investment
0.00      146
50.00       4
25.00       4
100.00      3
12.00       2
33.34       2
16.67       2
34.00       1
66.67       1
16.66       1
6.67        1
58.28       1
5.66        1
10.00       1
60.00       1
9.10        1
Name: count, dtype: int64


Unique Foreign Investment values:


Foreign_Investment
100.00    146
50.00       4
75.00       4
0.00        3
88.00       2
66.66       2
83.33       2
66.00       1
33.33       1
83.34       1
93.33       1
41.72       1
94.34       1
90.00       1
40.00       1
90.90       1
Name: count, dtype: int64

In [ ]:
#@title correlation matrix using only numerical columns
correlation_matrix = df4.select_dtypes(include=['number']).corr()

# a heatmap of the correlation matrix
fig = px.imshow(correlation_matrix,
                labels=dict(x="Features", y="Features", color="Correlation"),
                x=correlation_matrix.columns,
                y=correlation_matrix.columns,
                color_continuous_scale='RdBu',
                zmin=-1, zmax=1,
                title='Correlation Matrix of Numerical Features')
fig.show()

In [ ]:

display(df4.isna().sum())

INDUSTRY NAME                   0
DISTRICT                        0
TOTAL CAPITAL                   0
FIXED CAPITAL                   0
WORKING CAPITAL                 0
EMPLOYMENT                      0
CATEGORY                        0
SCALE                           0
POWER                           8
% OF INVESTMENT                 0
MONTH                           0
YEAR                            0
PROVINCE                        0
CATEGORY_FREQ                   0
SCALE_encode                    0
PROVINCE_BAGMATI                0
PROVINCE_GANDAKI                0
PROVINCE_KOSHI                  0
PROVINCE_LUMBINI                0
PROVINCE_MADHESH                0
PROVINCE_SUDUR-PASCHIM          0
TOTAL CAPITAL_standardized      0
FIXED CAPITAL_standardized      0
WORKING CAPITAL_standardized    0
Local_Investment                0
Foreign_Investment              0
dtype: int64

In [ ]:
#@title import train tesy split form scikit learn
from sklearn.model_selection import train_test_split

In [ ]:
#@title feature selection only relevant
features = ['FIXED CAPITAL_standardized', 'WORKING CAPITAL_standardized', 'POWER','SCALE_encode', 'Local_Investment', 'Foreign_Investment', 'PROVINCE_BAGMATI', 'PROVINCE_GANDAKI', 'PROVINCE_KOSHI', 'PROVINCE_LUMBINI', 'PROVINCE_MADHESH', 'PROVINCE_SUDUR-PASCHIM']
target = ['TOTAL CAPITAL_standardized']

In [ ]:
#@title define feature and target
X = df4[features]
y = df4[target]

In [ ]:
#@title split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
#@title import evalutation metrics
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:
#@title import random forest regressor
from sklearn.ensemble import RandomForestRegressor

In [ ]:
rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(X_train, y_train)
rf_predictions = rf_model.predict(X_test)
rf_mse = mean_squared_error(y_test, rf_predictions)
rf_r2 = r2_score(y_test, rf_predictions)

In [ ]:
#@title import xgboost
import xgboost as xgb

In [ ]:
xgb_model = xgb.XGBRegressor() # You can tune hyperparameters here
xgb_model.fit(X_train, y_train)
xgb_predictions = xgb_model.predict(X_test)
xgb_mse = mean_squared_error(y_test, xgb_predictions)
xgb_r2 = r2_score(y_test, xgb_predictions)

In [ ]:
#@title  print evaluation metrics in words

print("\nRandom Forest Regression:")
print(f"  Mean Squared Error: {rf_mse}")
print(f"  R-squared: {rf_r2}")

print("\nXGBoost Regression:")
print(f"  Mean Squared Error: {xgb_mse}")
print(f"  R-squared: {xgb_r2}")


Random Forest Regression:
  Mean Squared Error: 0.001293131809891337
  R-squared: 0.9983855849663145

XGBoost Regression:
  Mean Squared Error: 0.0062248479632699685
  R-squared: 0.9922285676002502


In [ ]:
#@title visual comparision
models = ['Random Forest', 'XGBoost']
mse_values = [rf_mse, xgb_mse]
r2_values = [ rf_r2, xgb_r2]

# subplots
fig = make_subplots(rows=1, cols=2, subplot_titles=('Mean Squared Error', 'R-squared'))

# MSE trace
fig.add_trace(go.Bar(x=models, y=mse_values, name='MSE'), row=1, col=1)
# R2 trace
fig.add_trace(go.Bar(x=models, y=r2_values, name='R2'), row=1, col=2)

# update layout
fig.update_layout(height=500, width=900, title_text="Comparison of MSE and R2 for Different Models")

# x-axis labels
fig.update_xaxes(title_text="Models", row=1, col=1)
fig.update_xaxes(title_text="Models", row=1, col=2)

# y-axis labels
fig.update_yaxes(title_text="MSE", row=1, col=1)
fig.update_yaxes(title_text="R2", row=1, col=2)

# plot
fig.show()

You've trained two models, Linear Regression and Random Forest Regression, to predict 'TOTAL CAPITAL'. The evaluation metrics you've provided are:

Mean Squared Error (MSE): Measures the average squared difference between predictions and actual values. Lower is better.
R-squared (R2): Represents the proportion of variance in the target variable explained by the model. Higher is better (ideally close to 1).
Here's a breakdown of the results:

Linear Regression:

Extremely Low MSE (3.235e-28): This indicates that the Linear Regression model's predictions are almost perfectly matching the actual 'TOTAL CAPITAL' values in the test data.
R-squared of 1.0: This means the model explains 100% of the variance in 'TOTAL CAPITAL'. It's a perfect score, suggesting a very strong linear relationship between the features and the target variable.
Random Forest Regression:

Higher MSE (0.00146): Compared to Linear Regression, Random Forest has a much higher MSE, indicating larger prediction errors.
High but slightly lower R-squared (0.998): While still very high, it's slightly lower than Linear Regression, suggesting that Random Forest might not be capturing the data patterns as effectively as Linear Regression in this case.
Conclusion and Potential Reasons:

Based on these metrics, Linear Regression is performing exceptionally well on your data. The near-perfect R-squared and extremely low MSE strongly suggest a linear relationship between the features you've selected and 'TOTAL CAPITAL'.

Random Forest, although a powerful model, might be overfitting slightly or not capturing the linear pattern as effectively as Linear Regression. This could be due to the hyperparameters needing more tuning or potential non-linearity in the data not being significant enough for Random Forest's advantages to take effect.

Recommendations:

Favor Linear Regression: The evidence strongly points to Linear Regression being more suitable for your specific data and prediction task.
Check for Overfitting: While Linear Regression seems to perform well, ensure it's not overfitting to the training data. Validate with more data if possible.
Feature Selection: Revisit your feature selection and selection process. It is possible that removing some features may improve linear model's performance even further.


encoding_methods = {
    'One-Hot Encoding': {
        'description': 'Creates binary (0 or 1) columns for each category in a categorical feature.',
        'best_for': 'Nominal categorical data (categories with no inherent order, e.g., color, country).  Avoids introducing ordinality.',
        'example': 'Converting "color" into separate columns like "color_red", "color_blue", etc.'
    },
    'Label Encoding': {
        'description': 'Assigns a unique integer to each category in a categorical feature.',
        'best_for': 'Ordinal categorical data (categories with a meaningful order, e.g., education level, size).',
        'example': 'Converting "size" (small, medium, large) into (0, 1, 2)'
    },
    'Target Encoding (Mean Encoding)': {
        'description': 'Replaces a categorical value with the average value of the target variable for that category.',
        'best_for': 'Categorical features with a strong relationship to the target variable. Prone to overfitting.',
        'example': 'Replace "city" with the average house price in that city.'
    },
    'Frequency Encoding': {
        'description': 'Replaces a category with its frequency (how many times it appears in the dataset).',
        'best_for': 'High cardinality categorical features (many unique values).  Captures the importance of frequent categories.',
        'example': 'Replace rare zip codes with their frequency.'
    },
    'Binary Encoding': {
        'description': 'Converts categories to binary code, representing each category as a sequence of bits.',
        'best_for': 'High cardinality features. Can be more efficient than one-hot encoding in terms of memory.',
        'example': 'Encode large number of countries'
    },
    'Ordinal Encoding': {
        'description': 'Similar to label encoding, assigns integer values based on order of categories.',
        'best_for': 'Ordinal categorical features, where the order matters. Custom mappings can be provided.',
        'example': 'Order categories: low, medium, high to numerical representations'
    },
     'Hashing Trick': {
        'description': 'Maps categorical values to a smaller set of indices using a hash function.',
        'best_for': 'Very high cardinality features, or when memory is a constraint.  Risk of collisions.',
        'example': 'Encode user IDs to a smaller dimensional space'
    }
}


Why Linear Regression has a higher R-squared and lower MSE:

Data Linearity: Linear Regression performs exceptionally well when the relationship between features and the target variable is linear. Your data might exhibit a strong linear pattern, allowing the model to capture it effectively, resulting in a near-perfect R-squared and a relatively low MSE.
Overfitting: While a very high R-squared is desirable, in some cases, it can be a sign of overfitting, especially if the model is too complex for the data. This means the model is fitting the training data too closely and might not generalize well to unseen data. However, in this case, the low MSE for Linear Regression indicates good performance on the data provided.
Feature Importance: Linear Regression assumes a linear relationship between features and target, making it less sensitive to outliers and irrelevant features. This can improve its performance on the current dataset.
Why Random Forest has a lower R-squared and higher MSE:

Data Complexity: Random Forest is a more complex model that can capture non-linear relationships in data. If the relationship between features and the target variable is not purely linear, it might struggle to fit the training data as precisely as Linear Regression, leading to a lower R-squared and higher MSE.
Hyperparameter Tuning: Random Forest has many hyperparameters (like n_estimators, max_depth) that can significantly affect its performance. Without proper tuning, the model might not be optimized for the dataset, resulting in suboptimal results.
Variance: Random Forest models, due to their ensemble nature, can have higher variance, potentially causing some overfitting. However, this model's overfitting should be less compared to linear regression.
In Summary

The higher R-squared and lower MSE suggest that Linear Regression is very likely the better model here for the data you've provided.
The improvement from hyperparameter tuning on the Random Forest is minimal, but you're at least reducing the risk of overfitting.
Recommendations:

Data Analysis: Dive deeper into your data with visualization or statistical analysis. This may reveal the true underlying relationship (linear vs non-linear) between features and the target.
Feature Selection: Carefully select features relevant to the target variable. It can reduce overfitting and improve the model's generalization ability.
Model Complexity: While complexity can handle non-linearity, too much complexity can lead to overfitting and a decrease in performance on unseen data.

Correlation Matrix and Heatmap: This helps you identify linear relationships between features and the target variable. Higher correlation values (closer to 1 or -1) indicate stronger linear relationships.
Scatter Plots: These plots visually show the relationship between individual features and the target. Look for patterns (linear, non-linear, clusters) to understand the nature of the relationship.
Pair Plots: Pair plots show relationships between multiple features simultaneously. They can reveal complex interactions and potential multicollinearity.
Statistical Analysis: Descriptive statistics provide insights into the distribution and central tendencies of your data. Hypothesis testing can help determine the statistical significance of relationships.

Select Numerical Features: Before calculating the correlation, we select only the numerical columns using df2.select_dtypes(include=['number']).columns and store them in the numerical_features variable.
Calculate Correlation for Numerical Features: We then apply the corr() function to a subset of the DataFrame containing only these numerical features: df2[numerical_features].corr(). This ensures that the function only operates on data that can be converted to float.
Updated Title: The title of the heatmap is changed to reflect that it is showing the correlation of "Numerical Features" for clarity.

Column Existence Check: I've added a check if feature in df2.columns: inside the loop. This ensures that the code only tries to create a scatter plot if the specified feature actually exists as a column in your DataFrame df2.
Warning Message: If a column is not found, a warning message is printed to inform you, and the loop skips to the next feature. This prevents the KeyError and allows the rest of your code to execute.

Box Plots: Box plots are created for each numerical feature to visually identify potential outliers (points outside the whiskers).
Scatter Plots with Outliers: Scatter plots are created for each feature against the target variable ('TOTAL CAPITAL'). Potential outliers, identified using the IQR method, are highlighted in red.
Cluster Analysis (KMeans):
The elbow method is used to determine the optimal number of clusters for KMeans.
KMeans clustering is applied, and a new 'Cluster' column is added to the DataFrame to store cluster assignments.
A scatter plot is used to visualize the clusters based on two features (you can modify the features used for visualization).

Select Numerical Features: Numerical features are selected from your DataFrame (df2) for PCA.
Standardization: The numerical features are standardized using StandardScaler to ensure that features with different scales do not dominate the PCA.
Apply PCA: PCA is applied using PCA from sklearn.decomposition. You need to specify the desired number of components (n_components).
Create DataFrame: A new DataFrame (df_pca) is created to store the principal components.
Add Cluster Assignments (Optional): If you have cluster assignments from previous analysis, you can add them to the df_pca DataFrame for visualization.
Visualization: A scatter plot is used to visualize the principal components, potentially colored by cluster assignments if available.

Feature Selection: Select the relevant features for clustering and create a subset of your DataFrame (X_cluster).
Standardization: Standardize the cluster features using StandardScaler to ensure that features with different scales do not disproportionately influence the clustering.
Elbow Method: Use the elbow method to determine the optimal number of clusters by visualizing the within-cluster sum of squares (WCSS) for different numbers of clusters. Choose the number of clusters where the WCSS starts to decrease less rapidly (the "elbow" point).
Apply KMeans: Apply KMeans clustering with the chosen number of clusters to your standardized data. Assign the cluster labels to a new column in your DataFrame (df2['Cluster']).
Analysis and Interpretation: Analyze and interpret the resulting clusters, for example, by calculating cluster statistics, visualizing the clusters, and identifying the characteristics of each cluster. This will help you understand the different segments in your data.

Feature Selection: Choose the relevant features for clustering.
Data Scaling: Standardize the features to ensure they have equal influence on the clustering.
Determine Optimal Number of Clusters (Elbow Method): Use the elbow method to visually identify the optimal number of clusters.
Apply KMeans Clustering: Apply KMeans with the chosen number of clusters to segment the data.
Analyze and Interpret Clusters: Analyze the characteristics of each cluster to understand the segments in your data. This might involve calculating cluster statistics, visualizing clusters using PCA or other techniques, and profiling clusters based on other features. Use these insights to develop strategies or gain a deeper understanding of your data.

Calculate Cluster Statistics:

df2.groupby('Cluster')[cluster_features].agg(['mean', 'std']) calculates the mean and standard deviation for each cluster feature within each cluster.
This provides a quantitative overview of the central tendency and variability within each cluster.
Profile Clusters based on Other Features:

The loop iterates through features like 'INDUSTRY NAME' and 'DISTRICT'.
df2.groupby(['Cluster', feature])['TOTAL CAPITAL'].agg(['mean', 'count']) calculates the mean 'TOTAL CAPITAL' and the count of observations for each combination of cluster and the selected feature.
This helps understand how clusters differ in terms of these categorical features.
Develop Strategies or Insights:

high_capital_clusters and low_capital_clusters are identified using the calculated cluster statistics.
This example highlights clusters with extreme values of 'TOTAL CAPITAL', which could be further investigated for potential strategies or insights.

Investigate clusters with specific industry or location characteristics.
Identify clusters with high variability in certain features, indicating potential heterogeneity within the cluster.
Develop targeted strategies for different clusters based on their unique characteristics.
Visualize the cluster profiles using bar plots or other appropriate visualizations.

# viz

In [ ]:
df0 = df4.copy()

In [ ]:
# @title Bar chart of CATEGORY counts
fig = px.histogram(df0, x="CATEGORY",
                   title="Distribution of Categories")
fig.show()

In [ ]:
# @title SCALE
fig = px.bar(df0.groupby('SCALE').size().reset_index(name='count'),
             x='SCALE', y='count',
             color='SCALE',
             title='Distribution of Scales')

fig.update_layout(
    xaxis_title='Scale',
    yaxis_title='Count',
    plot_bgcolor='rgba(0,0,0,0)',  # Set plot background to transparent
    xaxis={'showline': True, 'linewidth': 1, 'linecolor': 'black'},
    yaxis={'showline': True, 'linewidth': 1, 'linecolor': 'black'},
)
fig.update_traces(marker_line_width=1, marker_line_color='black') # Add border to bars
fig.show()

In [ ]:
# @title employment by district
df_employment = df0.groupby('CATEGORY')['EMPLOYMENT'].sum().sort_values(ascending=False)

fig = px.pie(df_employment, values='EMPLOYMENT', names=df_employment.index,
             title='Employment by Category')
fig.show()

In [ ]:
# @title stacked area chart
fig = px.bar(df0, x="CATEGORY", y="POWER",
             title="Power Consumption by Category")
fig.show()

In [ ]:
#@title power consumption by category
fig = px.pie(df0, values='POWER', names='CATEGORY', hole=0.4,
             title='Power Consumption by Category')
fig.show()

In [ ]:
#@title treemap total capital by category and scale
fig = px.treemap(df0, path=['SCALE','CATEGORY'], values='TOTAL CAPITAL',
                  title='Treemap of Total Capital by Category and Scale')
fig.show()

In [ ]:
#@title sunburst category scale employment
fig = px.sunburst(df0, path=['SCALE','CATEGORY'], values='EMPLOYMENT')
fig.show()

In [ ]:
# @title 3D Scatter Plot with Hover Information
fig = px.scatter_3d(df0, x='TOTAL CAPITAL', y='EMPLOYMENT', z='POWER', color='CATEGORY', hover_data=['DISTRICT', 'SCALE_encode'],
                    title='3D Scatter Plot with Hover Details')
fig.show()

In [ ]:
#@title Box Plots for Capital by Category
fig = px.box(df4, x='WORKING CAPITAL', y='CATEGORY', color='SCALE',
             title='Distribution of Total Capital by Category and Scale')
fig.show()

In [ ]:
# @title more advanced data analysis visualizations Funnel Chart
# Example using EMPLOYMENT and CATEGORY (Adjust values and labels)
employment_by_category = df0.groupby('CATEGORY')['EMPLOYMENT'].sum().reset_index()

fig = go.Figure(go.Funnel(
    y = employment_by_category['CATEGORY'],
    x = employment_by_category['EMPLOYMENT'],
    textposition = "inside",
    textinfo = "value+percent initial",
))
fig.update_layout(title="Funnel Chart of Employment by Category")
fig.show()


In [ ]:
# @title Parallel Categories Plot with more dimensions
fig = px.parallel_categories(df0, dimensions=['CATEGORY', 'DISTRICT', 'SCALE_encode'], color="EMPLOYMENT",
                             title='Parallel Categories Plot of Key Metrics')
fig.show()

In [ ]:
# @title Treemap Visualization
fig = px.treemap(df0, path=[ 'SCALE','CATEGORY', 'DISTRICT'], values='EMPLOYMENT',
                  color='TOTAL CAPITAL', hover_data=['POWER'],
                  color_continuous_scale='RdBu',
                  title="Treemap of Employment by Category, District, and Scale")
fig.show()

In [ ]:
# @title make visualization based on employment

fig = px.violin(df0, y="EMPLOYMENT", x="CATEGORY", color="CATEGORY", box=True, points="all",
                title="Violin Plot of Employment by Category")
fig.show()

In [ ]:
# @title Employment Generation per District:
district_summary = df0.groupby(['DISTRICT']).agg({'EMPLOYMENT': 'sum'})
district_summary = district_summary.sort_values(by='EMPLOYMENT', ascending=False)


# Show the DataFrame
display(district_summary)

# Visualize Employment by District
fig = px.bar(district_summary, x=district_summary.index, y='EMPLOYMENT',
             title='Employment by District')
fig.show()

EMPLOYMENT
DISTRICT                  
KATHMANDU             2763
LALITPUR              1527
KASKI                  462
DANG                   300
MORANG                 122
KAVRE                   90
GORKHA                  60
JHAPA                   50
PARSA                   45
SINDHUPALCHOWK          45
RUPANDEHI               41
CHITWAN                 40
SUNSARI                 13
BHAKTAPUR                9